dbms/src/Operators/HashJoinProbeTransformOp.cpp

```diff
@@ -100,39 +133,116 @@ OperatorStatus HashJoinProbeTransformOp::transformImpl(Block & block)
 {
     assert(status == ProbeStatus::PROBE);
     assert(probe_process_info.all_rows_joined_finish);
-    if likely (block)
-    {
-        join->checkTypes(block);
-        probe_process_info.resetBlock(std::move(block), 0);
-        assert(!probe_process_info.all_rows_joined_finish);
-    }
+    if (auto ret = probe_transform->tryFillProcessInfoInProbeStage(probe_process_info, block); ret != OperatorStatus::HAS_OUTPUT)
+        return ret;
 
     return onOutput(block);
 }
 
 OperatorStatus HashJoinProbeTransformOp::tryOutputImpl(Block & block)
 {
     if (status == ProbeStatus::PROBE && probe_process_info.all_rows_joined_finish)
-        return OperatorStatus::NEED_INPUT;
+    {
+        if (auto ret = probe_transform->tryFillProcessInfoInProbeStage(probe_process_info); ret != OperatorStatus::HAS_OUTPUT)
+            return ret;
+    }
 
     return onOutput(block);
 }
 
+void HashJoinProbeTransformOp::onWaitProbeFinishDone()
+{
+    if (probe_transform->needScanHashMapAfterProbe())
+    {
+        probe_transform->startScanHashMapAfterProbe();
+        switchStatus(ProbeStatus::READ_SCAN_HASH_MAP_DATA);
+    }
+    else if (probe_transform->shouldRestore())
+    {
+        switchStatus(ProbeStatus::GET_RESTORE_JOIN);
+    }
+    else
+    {
+        switchStatus(ProbeStatus::FINISHED);
+    }
+}
+
+void HashJoinProbeTransformOp::onRestoreBuildFinish()
+{
+    probe_transform->startRestoreProbe();
+    switchStatus(ProbeStatus::RESTORE_PROBE);
+}
+
+void HashJoinProbeTransformOp::onGetRestoreJoin()
+{
+    if (auto restore_exec = probe_transform->tryGetRestoreExec(); restore_exec)
+    {
+        probe_transform = restore_exec;
+        switchStatus(ProbeStatus::RESTORE_BUILD);
+    }
+    else
+    {
+        switchStatus(ProbeStatus::FINISHED);
+    }
+}
+
 OperatorStatus HashJoinProbeTransformOp::awaitImpl()
 {
-    if likely (status == ProbeStatus::WAIT_PROBE_FINISH)
+    while (true)
     {
-        if (join->isAllProbeFinished())
-        {
-            status = ProbeStatus::READ_SCAN_HASH_MAP_DATA;
-            scan_hash_map_after_probe_stream->readPrefix();
-            return OperatorStatus::HAS_OUTPUT;
-        }
-        else
+        switch (status)
         {
+        case ProbeStatus::WAIT_PROBE_FINISH:
+            if (probe_transform->quickCheckProbeFinished())
+            {
+                onWaitProbeFinishDone();
+                BREAK;
```

Comment:
Ditto

Reply:
Because it is in a while loop, here we check context.isCancelled in each loop of while to avoid canceling too slowly.